syntax = "proto3";
package payload.manager.core.v1;

import "google/protobuf/timestamp.proto";
import "payload/manager/core/v1/types.proto";
import "payload/manager/core/v1/policy.proto";
import "payload/manager/core/v1/id.proto";

/*
  Payload bytes are immutable after CommitPayload().
  Any modification MUST create a new PayloadID.
*/

message GpuLocation {
  uint32 device_id = 1;
  bytes ipc_handle = 2; // opaque CUDA IPC handle
  uint64 length_bytes = 3;
}

message RamLocation {
  string shm_name = 1;
  uint32 slab_id = 2;
  uint64 block_index = 3;
  uint64 length_bytes = 4;
}

message DiskLocation {
  string path = 1;
  uint64 offset_bytes = 2;
  uint64 length_bytes = 3;
}

/*
  Authoritative placement descriptor.

  version increments whenever:
    - tier changes
    - location changes
    - lifecycle state changes

  Clients MUST invalidate cached descriptors when version changes.
*/
message PayloadDescriptor {
  PayloadID payload_id = 1;
  Tier tier = 2;
  PayloadState state = 3;

  oneof location {
    GpuLocation gpu = 4;
    RamLocation ram = 5;
    DiskLocation disk = 6;
  }

  uint64 version = 7;
  google.protobuf.Timestamp created_at = 8;
  google.protobuf.Timestamp expires_at = 9;

  // advisory eviction behavior only
  EvictionPolicy eviction_policy = 10;
}
